### ISSUE TITLE
Plan for dropping/deprecating submodules of `py` and releasing v2.0

### ISSUE THREAD

**Issue by The-Compiler** [pytest-dev/py#288]
Prompted by #287, I'd propose to drop big parts of pylib in a (hopefully more or less final) py 2.0 release.

Clearly nobody is interested in maintaining pylib, the current situation seems to be that a couple of pytest maintainers keep it on life-support. While we already did a lot of work to [get rid of it in pytest](https://github.com/pytest-dev/pytest/issues/7259), this is still an ongoing process.

However, pytest [only uses](https://github.com/pytest-dev/pytest/blob/baada535a3e31ceb541a3e81319309a35bec934c/src/_pytest/compat.py#L45) `py.path`.

Everything else in pylib is being used by an small number of projects. Some of the submodules are both high-risk and almost zero-usage (`py.path.svn*`, looking at you!). Others have some real-world usage, but at some point, we should probably start deprecating those modules at least, so people can start looking for alternatives (given that pylib probably will cease to be maintained entirely after pytest is split off).

Thus, my proposal would be:

1) Either vendoring `py.path` in pytest (probably going to open a separate issue in the pytest repo for this), and then deprecating pylib in it's entirety, thus archiving this repository after dropping the `py` dependency in `pytest`.
2) Or dropping big parts of `py` (anything where we can't seem to find anything significant usage at the least, or perhaps even anything except `py.path`). This is what this issue will be about.

I used [Sourcegraph](https://sourcegraph.com/search) to try and get an overview over how the remaining parts of pylib are used. Here's what I found:

# `py.code`

Integrated into pytest with [pytest 2.9.0](https://docs.pytest.org/en/latest/changelog.html#id395) back in early 2016.

There's a [handful of remaining real-world usage](https://sourcegraph.com/search?q=context:global+%28%28from%7Cimport%29%5C+py%5C._%3Fcode%7Cfrom%5C+py%5C+import%5C+_%3Fcode%29+-file:py%28-%5B0-9.%5D*%29%3F/+-file:pytest/+-file:_pytest/+-repo:pytest-dev/py+-repo:cndn/intelligent-code-completion+-file:%5C.svn-base&patternType=regexp) (~30 results). The most striking one is probably [Intellij/PyDev monkeypatching it](https://github.com/JetBrains/intellij-community/blob/1330e199a4d1ec166efc8745a9cd2ca6f5af7284/python/helpers/pydev/_pydev_runfiles/pydev_runfiles_pytest2.py#L59-L69), but that does already account for the new `_pytest._code` location anyways. [Other than that](https://sourcegraph.com/search?q=context:global+%28%28from%7Cimport%29%5C+py%5C._%3Fcode%7Cfrom%5C+py%5C+import%5C+_%3Fcode%29+-file:py%28-%5B0-9.%5D*%29%3F/+-file:pytest/+-file:_pytest/+-repo:pytest-dev/py+-repo:cndn/intelligent-code-completion+-file:%5C.svn-base+-file:pydev_runfiles_pytest2%3F+-file:sourcetools%5C.py&patternType=regexp) I can see around 10 projects using it. Only 2 ([schevo](https://sourcegraph.com/github.com/Schevo/schevo/-/blob/schevo/lib/module.py), [pymtl3](https://sourcegraph.com/github.com/pymtl/pymtl3/-/blob/pymtl3/dsl/test/sim_utils.py?L95)) seem to be outside of testsuites. The rest are testsuites doing things like their own `pytest.raises`, which would probably be better off importing from `pytest._code` instead.

My verdict: +0.5 for dropping

# `py.error`

Used in [a couple of places](https://sourcegraph.com/search?q=context:global+%28%28from%7Cimport%29%5C+py%5C._%3Ferror%7Cfrom%5C+py%5C+import%5C+_%3Ferror%7C+py%5C._%3Ferror%5C.%29+-file:py%28-%5B0-9.%5D*%29%3F/+-file:pytest%28-%5B0-9.%5D*%29%3F/+-file:_pytest/+-repo:pytest-dev/py+-repo:cndn/intelligent-code-completion+-file:%5C.svn-base+-file:.*%5C.rst&patternType=regexp) (166 results), mostly out of pytest. From what I can see, mostly used in conjunction with `py.path`.

Superseded by [more fine-grained `OSError` exceptions](https://peps.python.org/pep-3151/) in Python 3. Sadly, `py.path.local` raises its own exceptions rather than those...

My verdict: -0.5 for dropping, should be kept as long as we keep `py.path` around for pytest. Thankfully, it's comparatively [small and simple](https://github.com/pytest-dev/py/blob/master/py/_error.py).

# `py.std`

[Deprecated](https://github.com/pytest-dev/py/commit/7f46178c972155f49ec5376b273f63492d9c7d2e) in 2017, [still some usage](https://sourcegraph.com/search?q=context:global+%28%28from%7Cimport%29%5C+py%5C._%3Fstd%7Cfrom%5C+py%5C+import%5C+_%3Fstd%7C+py%5C._%3Fstd%5C.%29+-file:py%28-%5B0-9.%5D*%29%3F/+-file:pytest%28-%5B0-9.%5D*%29%3F/+-file:_pytest/+-repo:pytest-dev/py+-repo:cndn/intelligent-code-completion+-file:%5C.svn-base+-file:.*%5C.rst&patternType=regexp), but from what I can tell, mostly in some old PyPy forks or something?

My verdict: Maybe needs some more investigation, but +1 for dropping, given it was deprecated before.

# `py.process`

[Imported in a couple of places](https://sourcegraph.com/search?q=context:global+%28%28from%7Cimport%29%5C+py%5C._%3Fprocess%7Cfrom%5C+py%5C+import%5C+_%3Fprocess%7C+py%5C._%3Fprocess%5C.%29+-file:py%28-%5B0-9.%5D*%29%3F/+-file:pytest%28-%5B0-9.%5D*%29%3F/+-file:_pytest/+-repo:pytest-dev/py+-repo:cndn/intelligent-code-completion+-file:%5C.svn-base+-file:.*%5C.rst&patternType=regexp). Funnily enough, the 3 or 4 projects I actually checked imported it, but never used it...

Maybe this happened due to some "automatic import" IDE feature for typo'd variables?

My verdict: +1 for dropping

# `py.apipkg`

[Zero usage](https://sourcegraph.com/search?q=context:global+%28%28from%7Cimport%29%5C+py%5C._%3Fapipkg%7Cfrom%5C+py%5C+import%5C+_%3Fapipkg%29+-file:py%28-%5B0-9.%5D*%29%3F/+-file:pytest/+-file:_pytest/+-repo:pytest-dev/py+-repo:cndn/intelligent-code-completion+-file:%5C.svn-basekg&patternType=regexp). Probably needs to be kept for pylib itself, but at least not exposed publicly.

# `py.iniconfig`

Split into [separate project](https://github.com/pytest-dev/iniconfig) for [pytest 6](https://docs.pytest.org/en/latest/changelog.html#pytest-6-0-0rc1-2020-07-08) in 2020.

Still [used in tox](https://github.com/tox-dev/tox/blob/d41bd622ea27b4a591be7aaf8f55560f5037b84f/src/tox/config/__init__.py#L1171) unfortunately, [1 other usage](https://sourcegraph.com/search?q=context:global+%28%28from%7Cimport%29%5C+py%5C._%3Finiconfig%7Cfrom%5C+py%5C+import%5C+_%3Finiconfig%29+-file:py%28-%5B0-9.%5D*%29%3F/+-file:pytest/+-file:_pytest/+-repo:pytest-dev/py+-repo:cndn/intelligent-code-completion+-file:%5C.svn-basekg&patternType=regexp) in [tox-factor](https://github.com/rpkilby/tox-factor/blob/bb36b281c8719af23c21ce78528f864bcdbd2e3e/src/tox_factor/test.py#L62).

My verdict: Probably wouldn't drop it right now, just because it's used in tox. We should coordinate with them to switch over to the split-off `iniconfig`, however. Maybe also find a way to raise deprecation warnings.

# `py.path.local`

Obviously keep. The only reason pylib is still around really...

# `py.path.svnwc`, `py.path.svnurl`, `py.path.SvnAuth`

The whole reason I'm opening this issue, see #287 and #256. Complex, high-risk, and just not worth any of the trouble.

Also, unused outside of what seems like [some PyPy development scripts](https://sourcegraph.com/search?q=context:global+py.path.svn+lang:python+-file:py/_path/+-file:/py%28-%5B0-9.%5D*%29%3F/testing/path/+-repo:pytest-dev/py+-file:pypy/tool/&patternType=standard), which are probably unused themselves, given that PyPy switched to Mercurial [12 years ago](https://www.pypy.org/posts/2010/12/pypy-migrates-to-mercurial-3308736161543832134.html).

My verdict: Burn it with fire.

# `py.builtin`

[Used in a surprisingly high number of places](https://sourcegraph.com/search?q=context:global+%28%28from%7Cimport%29%5C+py%5C._%3Fbuiltin%7Cfrom%5C+py%5C+import%5C+_%3Fbuiltin%7C+py%5C._%3Fbuiltin%5C.%29+-file:py%28-%5B0-9.%5D*%29%3F/+-file:pytest%28-%5B0-9.%5D*%29%3F/+-file:_pytest/+-repo:pytest-dev/py+-repo:cndn/intelligent-code-completion+-file:%5C.svn-base+-file:.*%5C.rst&patternType=regexp), including devpi and pytest plugins.

Probably mostly from "Python 2 support" times, but at the same time, it seems to be a lot of code churn for a [rather simple module](https://github.com/pytest-dev/py/blob/master/py/_builtin.py)

My verdict: -0.5 for removing it. Maybe find a way to raise deprecation warnings first.

# `py.io`

[Mostly used for `TerminalWriter`](https://sourcegraph.com/search?q=context:global+%28%28from%7Cimport%29%5C+py%5C._%3Fio%7Cfrom%5C+py%5C+import%5C+_%3Fio%7C+py%5C._%3Fio%5C.%29+-file:py%28-%5B0-9.%5D*%29%3F/+-file:pytest%28-%5B0-9.%5D*%29%3F/+-file:_pytest/+-repo:pytest-dev/py+-repo:cndn/intelligent-code-completion+-file:%5C.svn-base+-file:.*%5C.rst&patternType=regexp) in various projects from what it seems.

My verdict: -1 for removing it, as migration is probably more painful than e.g. for `py.builtin`. Might want to raise deprecation warnings telling people to use `rich` or something.

# `py.xml`

Apparently [used in various projects](https://sourcegraph.com/search?q=context:global+%28%28from%7Cimport%29%5C+py%5C._%3Fxml%7Cfrom%5C+py%5C+import%5C+_%3Fxml%7C+py%5C._%3Fxml%5C.%29+-file:py%28-%5B0-9.%5D*%29%3F/+-file:pytest%28-%5B0-9.%5D*%29%3F/+-file:_pytest/+-repo:pytest-dev/py+-repo:cndn/intelligent-code-completion+-file:%5C.svn-base+-file:.*%5C.rst&patternType=regexp), especially for HTML generation.

While it's high-risk, it's also arguably the most-used part of pylib (other than `py.path.local` and perhaps `TerminalWriter`). Maybe raise deprecation warnings? What alternatives are there?

# `py.log`

[Used in pypy and xdist](https://sourcegraph.com/search?q=context:global+%28%28from%7Cimport%29%5C+py%5C._%3Flog%7Cfrom%5C+py%5C+import%5C+_%3Flog%7C+py%5C._%3Flog%5C.%29+-file:py%28-%5B0-9.%5D*%29%3F/+-file:pytest%28-%5B0-9.%5D*%29%3F/+-file:_pytest/+-repo:pytest-dev/py+-repo:cndn/intelligent-code-completion+-file:%5C.svn-base+-file:.*%5C.rst&patternType=regexp) from what I can see.

As above: Probably deprecate, but do not remove.

---

That leaves us with:

- `py.code`: Drop, affected pytest monkeypatching code can switch to `pytest._code`
- `py.error`: Keep, useful for `py.path`
- `py.std`: Drop, long deprecated and moving away is trivial
- `py.process`: Drop (or possibly deprecate?)
- `py.apipkg`: Drop / make private
- `py.iniconfig`: Probably deprecate until tox moved away
- `py.path.local`: Keep obviously
- `py.path.svnwc`, `py.path.svnurl`, `py.path.SvnAuth`: Drop and kill it with fire
- `py.builtin`: Probably keep, maybe deprecate?
- `py.io`: Probably keep, maybe deprecate?
- `py.xml`: Probably keep, maybe deprecate?
- `py.log`: Probably keep, maybe deprecate?

@RonnyPfannschmidt @bluetech @nicoddemus what are your thoughts?

**Comment by nicoddemus** [comment]
Hi @The-Compiler,

Thanks a lot for the extensive research and detailed proposal!

Let me play devils advocate though: why the need to do this at all? I mean, another option is to just wait until we drop `py.path` from pytest, then officially archive the repository, without doing a new major release with removed modules, deprecation, etc.

**Comment by The-Compiler** [comment]
I see a couple of reasons:

1. As evident in #265, #256, #287, ..., people with no knowledge of the code or context around it seem to like reporting security "issues" against py, probably just because it's a popular library thanks to pytest. This wastes time all around - the reporter's, the maintainer's, and our user's. It also might give pytest a reputation of depending on "insecure" libraries if people who read those announcements then don't understand the context.
2. Dropping `py.path` from pytest will still take a while, with how entangled it is in the entire API. We've done a lot of progress, but sadly we aren't quite there yet. It now happened 2 or 3 times that people spent a lot of time worrying about code nobody should worry about. I have no doubts it's going to happen again until we actually manage to drop `py.path`. By removing code nobody is using, we reduce the likelyhood of that happening.
3. Even once 2. is done, I suspect we plan to have a `pytest-legacy-path` plugin or so. Given the ubiquity of `tmpdir`, I suspect it won't be too unpopular. It will need to either depend on pylib still, or vendor `py.path`, or try to provide a pathlib-based compatibility layer, but with `py.path`'s API. If it depends on pylib, I'm not entirely sure if we want to archive it. If it vendors it, why not do that now?
4. As seen in the searches above, there unfortunately *is* somewhat widespread pylib usage in the wild. This includes things like pytest plugins, projects somewhere in pytest's heritage (tox, devpi, PyPy, etc.), but also random other projects. We might not like that, and we might think of pylib only being around for pytest, but that's just not how things look. The minimum we can (and have to, IMHO) do is making all those users aware of the project being dead, and in the case of bigger projects (e.g. tox), maybe even help them migrate away. The best way to make them aware is probably a new release with deprecation warnings.

All in all, I feel like we need to have a proper deprecation plan for pylib before we get fully rid of it, whether we want to or not. I think dropping some ~unused code and deprecating some lightly used one would be a big first step towards that.

**Comment by RonnyPfannschmidt** [comment]
Hmm, I'd rather just put pylib in unmaintained state

**Comment by RonnyPfannschmidt** [comment]
In that case, apipkg has a own project, iniconfig as well, and time to drom/alias /deprecate

As for py. Io, I'd like to eventually separate it together with terminalwriter and pytests io capture, but the time horizon for that is too large

**Comment by nicoddemus** [comment]
I see, thanks @The-Compiler. Those are compelling arguments.

I think the way to go is to vendor `py.path` into pytest as you suggest, which allows pytest to drop the dependency to `py` entirely. We cannot deprecate `py.path.local` anytime soon, as old versions of pytest still use it. 

When the time comes for pytest to drop `py.local` support, the vendored code can be moved to the `pytest-legacy-path` repository as well.

About the other libraries, I don't see much point in dropping any of them, we should just deprecate everything except `py.path.local` and `py.error`: dropping would make sense if we ever intend for `py` to move forward without those libraries, but that is not the case, we plan to drop `py` entirely (archiving the repository) in the future.

**Comment by RonnyPfannschmidt** [comment]
@nicoddemus based on the cost of py.code vendoring and the to expect breakages of the api location changes i'm strictly opposed to vendoring py.path.local , id much rather see us move the support code into a temporaryly my plugin that will eventually turn a extra dependency and finally be phased out

**Comment by RonnyPfannschmidt** [comment]
my progression recommendation would be to
1. deprecate anything not in use in pytest/key pytest plugins
2. drop anything thats already extracted
3. remove svnwc as its broken and unmainained to begin with
4. completely deprecating pylib once pytest/key plugins are migrated off it

**Comment by asottile** [comment]
we could probably get away with vendoring and a little `sys.modules` trickery to preserve the module paths

**Comment by RonnyPfannschmidt** [comment]
i'd prefer to keep it outside the pytest codebase if feasible

**Comment by nicoddemus** [comment]
> @nicoddemus based on the cost of py.code vendoring and the to expect breakages of the api location changes i'm strictly opposed to vendoring py.path.local 

I don't think there's much *cost* involved in vendoring `py.path`, specially that with it we gain dropping  `py` from pytest, but I don't mind it much, and given you are strictly opposed, I won't argue in favor.

> my progression recommendation would be to

I agree, except with "dropping" anything: we really won't gain much by dropping stuff, given that we will just archive `py` in the future: dropping modules will just break working code needlessly, forcing some user to pin to an old version. But I'm in favor of deprecating everything of course (except `py.path` and `py.error`, as mentioned before), possibly mentioning alternatives that were extracted.

**Comment by The-Compiler** [comment]
> I agree, except with "dropping" anything: we really won't gain much by dropping stuff, given that we will just archive `py` in the future: dropping modules will just break working code needlessly, forcing some user to pin to an old version. But I'm in favor of deprecating everything of course (except `py.path` and `py.error`, as mentioned before), possibly mentioning alternatives that were extracted.

What we will gain is killing the noise of "security" advisories against pytest. This won't go away by deprecating things (we'd still be forced to fix it if we want people to not see security alerts for pytest). It will only go away by either dropping the `py` dependency from pytest (i.e. vendoring `py.path`), or by actually *removing* code nobody cares about, except apparently people looking for their next shiny CVE.

If we don't do either one or the other, the problem prompting me to open this issue won't be solved.

**Comment by RonnyPfannschmidt** [comment]
@nicoddemus the drop would bascially jsut add stuff like iniconfig and apipkg back to the dependency lsit instead of keeping a vendored copy

**Comment by asottile** [comment]
I put together a proof of concept: https://github.com/pytest-dev/pytest/pull/new/pylib-hax

**Comment by nicoddemus** [comment]
> What we will gain is killing the noise of "security" advisories against pytest. This won't go away by deprecating things (we'd still be forced to fix it if we want people to not see security alerts for pytest). It will only go away by either dropping the py dependency from pytest (i.e. vendoring py.path), or by actually removing code nobody cares about, except apparently people looking for their next shiny CVE.

OK, I agree. :+1:

**Comment by asottile** [comment]
ok so I've basically got the tests passing minus flake8 / mypy complaining -- the vendor hack actually doesn't seem that terrible? (it'll use `py` if installed, otherwise it'll use the vendored copy) -- this preserves the import paths -- https://github.com/pytest-dev/pytest/pull/10396

**Comment by kapilt** [comment]
re py.process - pytest-xdist depends pytest-forked which does use the module
https://github.com/pytest-dev/pytest-forked/blob/master/src/pytest_forked/__init__.py#L73

**Comment by asottile** [comment]
yep, we'll be removing `pytest-forked` as a dependency (as slated in the deprecation messages for the xdist 3.0.0 release)

**Comment by nicoddemus** [comment]
> we'll be removing pytest-forked as a dependency 

PR open: https://github.com/pytest-dev/pytest-xdist/pull/821 👍

**Comment by fschulze** [comment]
Regarding devpi:
1. I replaced several uses with recommended replacements today (``iniconfig``, ``py.builtin`` etc)
2. I'm unsure what to replace ``py.io.StdCaptureFD`` with. It is only used for tests, so I guess something (maybe internal) from pytest, but I'm unsure which of the capture stuff would be the best match. For subprocess calls, I can most likely replace it with ``check_output``, but there are other uses which would be harded. It is used in fixtures only afaict.
3. Once devpi-client drops Python 2.7 support (when it becomes work to fix running the tests) I can replace ``TerminalWriter`` with a wrapper for ``rich`` or something like that in ``devpi-common``.
4. I will most likely vendor ``py.path.local`` into ``devpi-common``, similar to pytest, but without the backward compatibility import stuff.
5. py.xml.html is already gone in next major ``devpi-web`` release branch.
6. I still have to investigate uses of ``py.error.E*``, might be related to vendoring ``py.path.local``.

**Comment by RonnyPfannschmidt** [comment]
I strongly recommend against vendoring of local path as long term solution, pytest intents to drop it as soon as possible

**Comment by bphunter1972** [comment]
FWIW, I use py.io.StdCaptureFD to capture output from a C process launched via pybind11 _which may hang_. I've found nothing else that could replace it, but I'm open to exploring alternatives. 

I'm willing to be wrong, but I don't think rich has anything like it.

**Comment by fschulze** [comment]
As replacement for StdCaptureFD I'm now using this:
```
from _pytest import capture

        cap = capture.MultiCapture(
            in_=capture.FDCapture(0),
            out=capture.FDCapture(1),
            err=capture.FDCapture(2))
```

**Comment by bphunter1972** [comment]
@fschulze That's all well and good if you happen to be using this for the purposes of pytest. 
I'd prefer a library having nothing to do with testing to capture output. That's what py.io provided.

**Comment by nicoddemus** [comment]
I personally do not know anything that implements `StdCapture` indeed.

An option might be for someone to fork `py.io` into its own library and maintain that separately; a smaller and more focused library might be an interesting little side project.

**Comment by fschulze** [comment]
Can anyone tell me what ``py.error.EBUSY`` is for and what exception is equivalent in Python? It seems to be Windows only and there is capture of it in devpi-server.

**Comment by nicoddemus** [comment]
I see this reference in the code:

https://github.com/pytest-dev/py/blob/6b219734fcd8d2f6489c2f50585a435b34c029c2/py/_error.py#L28

**Comment by sertonix** [comment]
Is importing `_pytest._io` an ok replacment for `py.io`? I am wondering if https://github.com/econchick/interrogate/pull/148 is a correct fix

**Comment by nicoddemus** [comment]
@sertonix you should keep using `py.io`, I think the consensus is to not deprecate it yet.

**Comment by sertonix** [comment]
Ok, thanks!
